tokio-timer/src/timer/mod.rs

```diff
@@ -0,0 +1,493 @@
+mod entry;
+mod handle;
+mod level;
+mod now;
+mod registration;
+
+use self::entry::Entry;
+use self::level::{Level, Expiration};
+
+pub use self::handle::{Handle, with_default};
+pub use self::now::{Now, SystemNow};
+pub(crate) use self::registration::Registration;
+
+use Error;
+use atomic::AtomicU64;
+
+use tokio_executor::park::{Park, Unpark, ParkThread};
+
+use std::{cmp, fmt};
+use std::time::{Duration, Instant};
+use std::sync::Arc;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::{SeqCst, Relaxed};
+use std::usize;
+
+/// Timer implementation that drives [`Sleep`], [`Interval`], and [`Deadline`].
+///
+/// # Implementation
+///
+/// `Timer` is based on the [paper by Varghese and Lauck][paper].
+///
+/// A hashed timing wheel is a vector of slots, where each slot handles a time
+/// slice. As time progresses, the timer walks over the slot for the current
+/// instant, and processes each entry for that slot. When the timer reaches the
+/// end of the wheel, it starts again at the beginning.
+///
+/// The `Timer` implementation maintains six wheels arranged in a set of levels.
+/// As the levels go up, the slots of the associated wheel represent larger
+/// intervals of time. At each level, the wheel has 64 slots. Each slot covers a
+/// range of time equal to the wheel at the lower level. At level zero, each
+/// slot represents one millisecond of time.
+///
+/// The wheels are:
+///
+/// * Level 0: 64 x 1 millisecond slots.
+/// * Level 1: 64 x 64 millisecond slots.
+/// * Level 2: 64 x ~4 second slots.
+/// * Level 3: 64 x ~4 minute slots.
+/// * Level 4: 64 x ~4 hour slots.
+/// * Level 5: 64 x ~12 day slots.
+///
+/// When the timer processes entries at level zero, it will notify all the
+/// [`Sleep`] instances as their deadlines have been reached. For all higher
+/// levels, all entries will be redistributed across the wheel at the next level
+/// down. Eventually, as time progresses, entries will `Sleep` instances will
+/// either be canceled (dropped) or their associated entries will reach level
+/// zero and be notified.
+///
+/// [`Sleep`]: ../struct.Sleep.html
+/// [`Interval`]: ../struct.Interval.html
+/// [`Deadline`]: ../struct.Deadline.html
+/// [paper]: http://www.cs.columbia.edu/~nahum/w6998/papers/ton97-timing-wheels.pdf
+#[derive(Debug)]
```

Comment:
Would you please make it `Clone`?

Reply:
What would `Clone` be useful for? It needs mutable access to itself to process entries, so you couldn't really be polling it in multiple places at once.

Reply:
Timer cannot be clone as the `Timer` value is what maintains all the timer state. `Handle` is how you create timeouts and this type is `Clone`.

Reply:
Al right. Would you provide an example of how to create a 1 instance of timer only and set timeout for `io::copy` right [here](https://github.com/tokio-rs/tokio/blob/master/examples/echo.rs#L79)?

Reply:
Once this is fully integrated, you would simply do `my_fut.deadline(instant)`